DietPi-Software | General internal cleanup of the choice system

MichaIng opened the pull request:
**Status**: Ready

**Commit list/description**:
+ DietPi-Software | General internal cleanup of the choice system

SSH Server and File Server choices have no index variable anymore, but the choice is derived from the actual install state of the related software titles. ~Also multiple file servers can now be selected, the menu has changed to a checkbox as it may make much sense to run different file servers concurrently for different purpose, e.g. LAN and WAN file transfer, or to support different client operating systems. Consequently, all file servers have been added to the checkbox dialog.~

Install and uninstall marks are now worked out in a more generic consistent way. The `CHOICEMADE` variables have consequently gone, replaced by the loop which collects the actually installed and uninstalled software titles, that existed before, just after `CHOICEMADE` evaluation.

**ToDo**:
- [x] Prevent concurrent ProFTPD and vsftpd installs. E.g. deselect vsftpd automatically in case, as part of `Mark_Dependencies()` where we handle other conflicts, like Pi-hole + AdGuard Home?
- [x] Remove the `UNINSTALL_REQUIRED` variable, instead check for actual `-1` marks to call the uninstall function after installs.
- [x] If due to choice changes only uninstalls are done, and it is not the first run setup, skip the install function and call the uninstall function directly.
- [x] Adjust choice and preference indices automatically on individual software installs and uninstalls. E.g. when DietPi-RAMlog is uninstalled, switch to logging index `0` or `-3` automatically, depending on whether Rsyslog is installed or not. And when Chromium is installed, apply it as browser preference automatically.
- [x] Apply preference changes immediately. So when exiting `dietpi-software`, the selection remains. Also `CURRENT` and `TARGET` variables can be merged.
- [x] Block preference changes in general when a related software is installed already? Hence replace the menu with a message that the preference has no effect anymore because software X is installed already. Practically preferences are already ignored in such case, so it makes sense inform users about this in the first place to prevent confusion when Firefox is selected as browser preference but doesn't get installed because Chromium already was. The above task then additionally makes sense so that the "blocked" preference shows what is installed already.
- [x] Evaluate whether it makes sense to keep the file server index variable in `dietpi.txt`, or whether to replace it with the already existing option to add software IDs for automatic install. A practical difference is that the `AUTO_SETUP_INSTALL_SOFTWARE_ID`'s only have an effect on `AUTO_SETUP_AUTOMATED=1` installs, while the `AUTO_SETUP_FILE_SERVER_INDEX` pre-sets the related install mark as well for the interactive menu. So one does not need to browse/search the software list anymore but can hit `Install` directly.
- [x] Resolve conflicts in a consistent way: https://github.com/MichaIng/DietPi/commit/dee14647073f231c7a97200d5b7505b1ffa43abb#commitcomment-60697523

MichaIng committed:
v7.9

+ DietPi-Software | General internal cleanup of the choice system

MichaIng committed:
v7.9

+ DietPi-Software | General internal cleanup of the choice system

MichaIng committed:
v7.9

+ DietPi-Software | Don't let ProFTPD and vsftpd be installed at the same time. When selected via file server choice menu, deselect ProFTPD if vsftpd is installed already, else unmark vsftpd.
+ DietPi-Software | Generally when unmarking conflicting software, do not unmark anything or show a warning if the conflicting software titles got installed already. Effectively nothing is unmarked then, and it shouldn't so that uninstalls can still be done.

MichaIng committed:
v7.9

+ DietPi-Software | Don't let ProFTPD and vsftpd be installed at the same time. When selected via file server choice menu, deselect ProFTPD if vsftpd is installed already, else unmark vsftpd.
+ DietPi-Software | Generally when unmarking conflicting software, do not unmark anything or show a warning if the conflicting software titles got installed already. Effectively nothing is unmarked then, and it shouldn't so that uninstalls can still be done.

MichaIng committed:
v7.9

+ DietPi-Software | Remove UNINSTALL_REQUIRED variable and instead check for any actual "-1" mark to trigger the Uninstall_Software function call for CLI and after installs. Also the uninstall menu code has been simplified, getting rid of of another flag variable in favour of checking whether any software has been added to the anyway created G_WHIP array.

MichaIng committed:
v7.9

+ DietPi-Software | Remove UNINSTALL_REQUIRED variable and instead check for any actual "-1" mark to trigger the Uninstall_Software function call for CLI and after installs. Also the uninstall menu code has been simplified, getting rid of of another flag variable in favour of checking whether any software has been added to the anyway created G_WHIP array.

MichaIng committed:
v7.9

+ DietPi-Software | If due to choice changes only uninstalls are done and it is not the first run setup, skip the install function and call the uninstall function directly.

MichaIng committed:
v7.9

+ DietPi-Software | If due to choice changes only uninstalls are done and it is not the first run setup, skip the install function and call the uninstall function directly.

MichaIng committed:
v7.9

+ DietPi-Software | Remove obsolete tabs

MichaIng committed:
v7.9

+ DietPi-Software | Remove obsolete tabs

MichaIng committed:
v7.9

+ DietPi-Software | Remove abortion banner: No need to erase the console, especially when no changes have been applied. Also the function contained a 1st run abort (short version) banner which was never used, as aborting DietPi-Software on first run setup was prevented in the first place, reasonably, to have all setup steps done regardless of actual software selections.

MichaIng committed:
v7.9

+ DietPi-Software | Remove abortion banner: No need to erase the console, especially when no changes have been applied. Also the function contained a 1st run abort (short version) banner which was never used, as aborting DietPi-Software on first run setup was prevented in the first place, reasonably, to have all setup steps done regardless of actual software selections.

MichaIng committed:
v7.9

+ DietPi-Software | Adjust logging choice index on Rsyslog and DietPi-RAMlog installs and uninstalls. Generally assure that RAMlog index stays active as long as /var/log is still a tmpfs, to rule out filled tmpfs until reboot. On DietPi-RAMlog uninstall, the logging index will be adjusted by the service which does the actual disabling on next reboot.

MichaIng committed:
v7.9

+ DietPi-Software | Adjust logging choice index on Rsyslog and DietPi-RAMlog installs and uninstalls. Generally assure that RAMlog index stays active as long as /var/log is still a tmpfs, to rule out filled tmpfs until reboot. On DietPi-RAMlog uninstall, the logging index will be adjusted by the service which does the actual disabling on next reboot.

MichaIng committed:
v7.9

+ DietPi-Software | Merge webserver preference index variables, prevent preference choice if a webserver is installed already, apply choices immediately (persistent without doing an actual isntall) and adjust index on webserver uninstalls.

MichaIng committed:
v7.9

+ DietPi-Software | Merge webserver preference index variables, prevent preference choice if a webserver is installed already, apply choices immediately (persistent without doing an actual isntall) and adjust index on webserver uninstalls.

MichaIng committed:
v7.9

+ DietPi-Software | Merge desktop preference index variables, prevent preference choice if a desktop is installed already, apply choices immediately (persistent without doing an actual isntall) and adjust index on desktop uninstalls.

MichaIng committed:
v7.9

+ DietPi-Software | Merge desktop preference index variables, prevent preference choice if a desktop is installed already, apply choices immediately (persistent without doing an actual isntall) and adjust index on desktop uninstalls.

MichaIng committed:
v7.9

+ DietPi-Software | Merge browser preference index variables, prevent preference choice if a browser is installed already, apply choices immediately (persistent without doing an actual isntall) and adjust index on browser uninstalls.

MichaIng committed:
v7.9

+ DietPi-Software | Merge browser preference index variables, prevent preference choice if a browser is installed already, apply choices immediately (persistent without doing an actual isntall) and adjust index on browser uninstalls.

MichaIng committed:
v7.9

+ DietPi-Software | Remove file server choice setting from dietpi.txt, but keep code to apply the choice to be backwards compatible with current/old images. Generally there is not much reason so give file servers any special place over any other software that allows to transfer files to/from the server (any we have many of them). Those can be added for automated install respectively selected via "Browse software" or "Search software" and we may hence remove the file server choice menu in the future.

MichaIng committed:
v7.9

+ DietPi-Software | Remove file server choice setting from dietpi.txt, but keep code to apply the choice to be backwards compatible with current/old images. Generally there is not much reason so give file servers any special place over any other software that allows to transfer files to/from the server (any we have many of them). Those can be added for automated install respectively selected via "Browse software" or "Search software" and we may hence remove the file server choice menu in the future.

MichaIng committed:
v7.9

+ DietPi-Software | Remove the File Server selection menu. There is no reason to treat file servers any different than other software titles, they are no dependencies, do not generally conflict with each other and do not affect other installs.

MichaIng committed:
v7.9

+ DietPi-Software | Remove the File Server selection menu. There is no reason to treat file servers any different than other software titles, they are no dependencies, do not generally conflict with each other and do not affect other installs.

MichaIng committed:
v7.9

+ DietPi-Software | Fallback to correct default when resolving dependencies. This is done already in parent functions but it doesn't hurt to do this as well in the final function.

MichaIng committed:
v7.9

+ DietPi-Software | Fallback to correct default when resolving dependencies. This is done already in parent functions but it doesn't hurt to do this as well in the final function.

MichaIng committed:
v7.9

+ DietPi-Software | Update webserver stack meta install states right before writing the installed states file. This assures that those always represent the actual webserver and database install states after every install or uninstall.
+ DietPi-Software | Implement a new conflict system which allows to consistently unmark and warn users about conflicting software selections, when doing this interactively from the menus as well as when installing via CLI
+ DietPi-Software | Reorder install steps so that unmarks are done first, and the install is aborted if no marks are left, and required directories only after successfuly network check + time sync + APT
+ DietPi-Software | Do not warn or recommend to leave webserver stacks unmarked to be auto-selected as dependency. There is absolutely no problem when users do this manually. It is not required (as long as the webserver stack is not wanted as standalone installation) but the additional step has been done already, so the warning is pointless. The new conflict system additionally prevents and warns about conflicts, if this was one of the reasons behind the warning.

MichaIng committed:
v7.9

+ DietPi-Software | Update webserver stack meta install states right before writing the installed states file. This assures that those always represent the actual webserver and database install states after every install or uninstall.
+ DietPi-Software | Implement a new conflict system which allows to consistently unmark and warn users about conflicting software selections, when doing this interactively from the menus as well as when installing via CLI
+ DietPi-Software | Reorder install steps so that unmarks are done first, and the install is aborted if no marks are left, and required directories only after successfuly network check + time sync + APT
+ DietPi-Software | Do not warn or recommend to leave webserver stacks unmarked to be auto-selected as dependency. There is absolutely no problem when users do this manually. It is not required (as long as the webserver stack is not wanted as standalone installation) but the additional step has been done already, so the warning is pointless. The new conflict system additionally prevents and warns about conflicts, if this was one of the reasons behind the warning.

MichaIng committed:
v7.9

+ DietPi-Software | Remove the flag for software which requires automated APT upgrades. Keeping APT packages updates is generally and always recommended, if maintenance can be efforted and there is no reason why this would be more important when using a VPN server or Tor compared to e.g. hosting personal data via webserver or similar.

MichaIng committed:
v7.9

+ DietPi-Software | Remove the flag for software which requires automated APT upgrades. Keeping APT packages updates is generally and always recommended, if maintenance can be efforted and there is no reason why this would be more important when using a VPN server or Tor compared to e.g. hosting personal data via webserver or similar.

MichaIng committed:
v7.9

+ DietPi-Software | Remove extra newline in .installed state file
+ DietPi-Software | Consequently loop through aSOFTWARE_NAME instead of aSOFTWARE_INSTALL_STATE keys to not handle install states of software which is not defined anymore, e.g. a forgotten install state while the related software title has been removed with a recent DietPi release. There is an install state assigned in every case as part of the array creation.
+ DietPi-Software | Do not show webserver stacks in uninstall menu: Their install states will be aligned with webserver/database install states automatically.

MichaIng committed:
v7.9

+ DietPi-Software | Remove extra newline in .installed state file
+ DietPi-Software | Consequently loop through aSOFTWARE_NAME instead of aSOFTWARE_INSTALL_STATE keys to not handle install states of software which is not defined anymore, e.g. a forgotten install state while the related software title has been removed with a recent DietPi release. There is an install state assigned in every case as part of the array creation.
+ DietPi-Software | Do not show webserver stacks in uninstall menu: Their install states will be aligned with webserver/database install states automatically.

MichaIng committed:
v7.9

+ DietPi-Patches | Remove obsolete SSH server and file server indices as well

MichaIng committed:
v7.9

+ DietPi-Patches | Remove obsolete SSH server and file server indices as well

MichaIng committed:
v7.9

+ DietPi-Software | Fix syntax and shorten DietPi-Dashboard description to not blow checklist dimensions

MichaIng committed:
v7.9

+ DietPi-Software | Fix syntax and shorten DietPi-Dashboard description to not blow checklist dimensions

MichaIng committed:
v7.9

+ DietPi-Software | Syntax

MichaIng committed:
v7.9

+ DietPi-Software | Syntax

MichaIng committed:
v7.9

+ DietPi-Set_software | Assure that systemd-timesyncd is installed on Bullseye and up, when trying to enable and/or start it. And only try to stop and disable the service when it is either enabled or running.

MichaIng committed:
v7.9

+ DietPi-Drive_Manager | When checking for Btrfs kernel module, do load. It is required shortly after.

Joulinar commented on dietpi/dietpi-software:
```diff
@@ -1438,6 +1436,7 @@ INDEX_BROWSER_TARGET=$INDEX_BROWSER_TARGET"
 		aSOFTWARE_CATX[$software_id]=-1
 		aSOFTWARE_DOCS[$software_id]='https://dietpi.com/docs/software/webserver_stack/#nginx'
 		aSOFTWARE_DEPS[$software_id]='89'
+		aSOFTWARE_CONFLICTS[$software_id]='75 76 81 82 83 85'
```
I guess it should be `75 76 81 82 83 84` as ID `85` is Nginx itself

MichaIng replied:
Good catch 👍.

Joulinar commented on dietpi/dietpi-software:
```diff
@@ -1406,6 +1400,7 @@ INDEX_BROWSER_TARGET=$INDEX_BROWSER_TARGET"
 		aSOFTWARE_CATX[$software_id]=12
 		aSOFTWARE_DOCS[$software_id]='https://dietpi.com/docs/software/webserver_stack/#llsp-web-stack'
 		aSOFTWARE_DEPS[$software_id]='84 87 89'
+		aSOFTWARE_CONFLICTS[$software_id]='75 76 78 79 83 85'
```
should be `83 84`

MichaIng replied:
🤔 There are quite some things I don't like about GitHub desktop: You cannot simply rebase and then add a commit and push. It sees a conflicts between the commit you just did and every other commit that was not pushed yet. Then when you solve the conflicts, it merges base branch changes a way that they still show up as changes, even that they are present and were even coming from the base branch. So now we have unrelated changes shown in this PR... Furthermore, even that your three code comments were address, two of them do not show up as "outdated". More confusingly, they are attached to a line that looks very similar than before, if indeed the problem were not addressed, but it's the line on Lighttpd, hence it somehow changed the location of your comment in a way that can be dangerously misinterpreted...

And worst about it is that there is no way to revert this mess, once pushed. The merge commit to resolve the conflicts does not show any changes itself, even that it lead to the additional changes from the base branch now erroneously shown here. Trying to undo the commit hence gives `[!] There are no changes to commit.`.

At least I convinced GitHub desktop developers to allow amending to commits which were pushed already. But still while using GitHub desktop I regularly run into situations where it denies something that works perfectly fine with `git` CLI, where some step takes ages, respectively literally never finishes (also reported upstream, not only by me), because it seems to always try to stash/trashbin every step in a way that takes ages on larger repos, instead of allowing to simply reset to where you want, and I regularly fear that something is broken 😢.

Joulinar commented on dietpi/dietpi-software:
```diff
@@ -1414,6 +1409,7 @@ INDEX_BROWSER_TARGET=$INDEX_BROWSER_TARGET"
 		aSOFTWARE_CATX[$software_id]=12
 		aSOFTWARE_DOCS[$software_id]='https://dietpi.com/docs/software/webserver_stack/#llmp-web-stack'
 		aSOFTWARE_DEPS[$software_id]='84 88 89'
+		aSOFTWARE_CONFLICTS[$software_id]='75 76 78 79 83 85'
```
should be `83 84`

MichaIng committed:
v7.9

+ DietPi-Software | Fix Nginx conflicting with itself

MichaIng committed:
Merge branch 'choices' of https://github.com/MichaIng/DietPi into choices

MichaIng committed:
v7.9

+ DietPi-Software | Fix false software title shown in conflicts prompt when a conflict was installed already

MichaIng committed:
v7.9

+ DietPi-Software | Whoopsie: Check for conflicts on subsequent menu selections

MichaIng committed:
v7.9

+ DietPi-Software | Fix code comment as well

MichaIng committed:
v7.9

+ DietPi-Software | Tiny

MichaIng committed:
v7.9

+ DietPi-Software | Add a dependant loop when unmarking conflicts, so that all dependants and in case their dependants are unmarked as well. Currently there is no actual case where this is required, but it doesn't hurt to be failsafe.

MichaIng committed:
v7.9

+ DietPi-Software | When using a CLI reinstall, skip the conflicts loop to assure that no software is marked as uninstalled which was previously installed already. As the conflict system is new, it is possible that conflicting software is installed already, and probably it is well configured, like one webserver as proxy on port 443 and another one internally on a different port. As a general rule: We do not unmark software which is installed already, and the CLI reinstall can only reinstall software that is installed already.
+ DietPi-Software | Tune preference system texts to be hopefully better understandable: https://github.com/MichaIng/DietPi/issues/4914#issuecomment-955636214

MichaIng committed:
v7.9

+ DietPi-Software | Syntax
